### PR TITLE
Pimcore requires doctrine/dbal ^2.11.0 and php ^7.3

### PR DESCRIPTION
**Pull request by blankse**
The PR https://github.com/pimcore/pimcore/pull/7327 use \Doctrine\DBAL\Exception which is first created in doctrine/dbal 2.11.0.
doctrine/dbal 2.11.0 requires php ^7.3

### CODE REVIEW

**Comment by brusch** [comment]
Just wanted to comment that Travis needs to be changed 😉 Thanks for updating the PR.

**Comment by brusch** [comment]
Related to #7410
@blankse can you maybe also fix the deprecations in this PR?

**Comment by jdreesen** [comment]
`doctrine/dbal v2.11` has many more deprecations that should also be fixed: https://github.com/doctrine/dbal/blob/2.12.x/UPGRADE.md#upgrade-to-211

**Comment by blankse** [comment]
@brusch @jdreesen I fixed many 2.11 deprecations.

There are deprecations in QueryBuilder and a deprecation with the getParams() (is internal now).
QueryBuilder is also deprecated, so I ignore it for now.
For getParams exists no alternative. We should load it direct from the settings?

But we can leave it at that for now, or?

**Comment by brusch** [comment]
@blankse nice, thanks a lot! 
`QueryBuilder`  👍
`getParams()` yep, we have to get the info then from the settings - are you updating the PR? 😊

**Comment by brusch** [comment]
@blankse regarding `>=` PHP 7.3 we should update this in the docs as well: 
https://github.com/pimcore/pimcore/blob/ae7b62cc4a0ade1bac5ad70d720187215344f827/doc/Development_Documentation/23_Installation_and_Upgrade/01_System_Requirements.md#L12

**Comment by blankse** [comment]
@brusch I looked in the container. There are no symfony parameter for the db settings. They are only set to the Connection cunstructur.

**Comment by brusch** [comment]
@blankse hm, too bad, then we need to put it on the container as a parameter I guess, probably here is a good place for that: 
https://github.com/pimcore/pimcore/blob/9af9958fbcfb9e86406f8076212196a52b831b81/bundles/CoreBundle/DependencyInjection/PimcoreCoreExtension.php#L70
wdyt?

**Comment by blankse** [comment]
@brusch From where do you want get the settings in the PimcoreCoreExtension? The config variable have only the settings from the pimcore configuration tree.

The db settings are in the doctrine configuration tree.

**Comment by brusch** [comment]
@blankse 🤦‍♂️ you're right, sorry 😊

**Comment by brusch** [comment]
I think using the `@internal` tagged `getParams()` is fine for now, if Doctrine changes something there in the future we can always add our own getter to `\Pimcore\Db\PimcoreExtensionsTrait`.

**Comment by blankse** [comment]
@brusch I agree with you

File: lib/Db/PimcoreExtensionsTrait.php [hunk 1]
```diff
@@ -14,7 +14,12 @@
 
 namespace Pimcore\Db;
 
+use Doctrine\DBAL\Cache\CacheException;
 use Doctrine\DBAL\Cache\QueryCacheProfile;
+use Doctrine\DBAL\Driver\Exception as DriverException;
+use Doctrine\DBAL\Driver\PDO\Statement as PDOStatement;
```
Comment (brusch) [inline lib/Db/PimcoreExtensionsTrait.php:20]:
```suggestion
use Doctrine\DBAL\Driver\Result;
```

File: lib/Db/PimcoreExtensionsTrait.php [hunk 2]
```diff
@@ -278,12 +288,13 @@ public function fetchCol($sql, $params = [], $types = [])
         // unfortunately Mysqli driver doesn't support \PDO::FETCH_COLUMN, so we have to do it manually
         $stmt = $this->executeQuery($sql, $params, $types);
         $data = [];
-        while (($row = $stmt->fetchColumn()) || $row !== false) {
-            $data[] = $row;
+        if ($stmt instanceof PDOStatement) {
```
Comment (brusch) [inline lib/Db/PimcoreExtensionsTrait.php:291]:
```suggestion
        if ($stmt instanceof Result) {
```

File: lib/Db/PimcoreExtensionsTrait.php [hunk 3]
```diff
@@ -315,15 +328,18 @@ public function fetchOne($sql, $params = [], $types = [])
...
      */
     public function fetchPairs($sql, array $params = [], $types = [])
     {
         $params = $this->prepareParams($params);
-        $statement = $this->executeQuery($sql, $params, $types);
+        $stmt = $this->executeQuery($sql, $params, $types);
         $data = [];
-        while ($row = $statement->fetch(\PDO::FETCH_NUM)) {
-            $data[$row[0]] = $row[1];
+        if ($stmt instanceof PDOStatement) {
```
Comment (brusch) [inline lib/Db/PimcoreExtensionsTrait.php:339]:
```suggestion
        if ($stmt instanceof Result) {
```

File: lib/Targeting/Storage/DbStorage.php [hunk 4]
```diff
@@ -18,8 +18,9 @@
 namespace Pimcore\Targeting\Storage;
 
 use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Driver\PDO\Statement as PDOStatement;
```
Comment (brusch) [inline lib/Targeting/Storage/DbStorage.php:21]:
```suggestion
use Doctrine\DBAL\Driver\Result;
```

File: lib/Targeting/Storage/DbStorage.php [hunk 5]
```diff
@@ -86,11 +87,12 @@ public function all(VisitorInfo $visitorInfo, string $scope): array
         $this->addExpiryParam($qb, $scope);
 
         $stmt = $qb->execute();
-        $result = $stmt->fetchAll();
-
         $data = [];
-        foreach ($result as $row) {
-            $data[$row['name']] = json_decode($row['value'], true);
+
+        if ($stmt instanceof PDOStatement) {
```
Comment (brusch) [inline lib/Targeting/Storage/DbStorage.php:92]:
```suggestion
        if ($stmt instanceof Result) {
```

File: lib/Targeting/Storage/DbStorage.php [hunk 6]
```diff
@@ -119,7 +121,11 @@ public function has(VisitorInfo $visitorInfo, string $scope, string $name): bool
         $this->addExpiryParam($qb, $scope);
 
         $stmt = $qb->execute();
-        $result = (int)$stmt->fetchColumn();
+        $result = 0;
+
+        if ($stmt instanceof PDOStatement) {
```
Comment (brusch) [inline lib/Targeting/Storage/DbStorage.php:126]:
```suggestion
        if ($stmt instanceof Result) {
```

File: lib/Targeting/Storage/DbStorage.php [hunk 7]
```diff
@@ -177,7 +183,11 @@ public function get(VisitorInfo $visitorInfo, string $scope, string $name, $defa
         $this->addExpiryParam($qb, $scope);
 
         $stmt = $qb->execute();
-        $result = $stmt->fetchColumn();
+        $result = false;
+
+        if ($stmt instanceof PDOStatement) {
```
Comment (brusch) [inline lib/Targeting/Storage/DbStorage.php:188]:
```suggestion
        if ($stmt instanceof Result) {
```

File: lib/Targeting/Storage/DbStorage.php [hunk 8]
```diff
@@ -292,9 +302,12 @@ private function loadDate(VisitorInfo $visitorInfo, string $scope, string $selec
         $this->addExpiryParam($qb, $scope);
 
         $stmt = $qb->execute();
-        $date = $this->convertToDateTime($stmt->fetchColumn());
 
-        return $date;
+        if ($stmt instanceof PDOStatement) {
```
Comment (brusch) [inline lib/Targeting/Storage/DbStorage.php:306]:
```suggestion
        if ($stmt instanceof Result) {
```

**Review by brusch — CHANGES_REQUESTED** [review]
I think it's more generic if we use `\Doctrine\DBAL\Driver\Result` instead of `\Doctrine\DBAL\Driver\PDO\Statement` then `Mysqli` is still an option 😊

**Pull request merged**